sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/transforms/Combine.java

```diff
@@ -269,9 +282,22 @@ private Combine() {
    */
   public static <K, InputT, OutputT> GroupedValues<K, InputT, OutputT> groupedValues(
       PerKeyCombineFn<? super K, ? super InputT, ?, OutputT> fn) {
-    return new GroupedValues<>(fn);
+    return groupedValues(fn, getClassSafe(fn));
   }
 
+  private static <K, InputT, OutputT> GroupedValues<K, InputT, OutputT> groupedValues(
+          PerKeyCombineFn<? super K, ? super InputT, ?, OutputT> fn, Class<?> fnClass) {
+    return new GroupedValues<>(fn, fnClass);
+  }
+
+  private static Class<?> getClassSafe(Object fn) {
```

Comment:
I think that this should be baked into the `add` logic for classes, since otherwise we'll need to take care to do this in all future cases. I'm surprised that serialization breaks display data -- when do we ever serialize them? Also, could we fall back on using the classes toString?

If this is truly just about serialization, make the field `transient` (so its not serialized) since we only need to keep it down to the point where we extract display data. Or, put it inside the display data function?